src/components/selection_input/selection_input.ts

```diff
@@ -245,19 +241,74 @@ export class SelectionInput extends Component<Props, SpreadsheetChildEnv> {
     this.props.onSelectionConfirmed?.();
   }
 
-  disable() {
+  confirm() {
     this.env.model.dispatch("UNFOCUS_SELECTION_INPUT");
-    const ranges = this.env.model.getters.getSelectionInputValue(this.id);
+    const ranges = this.env.model.getters
+      .getSelectionInput(this.id)
+      .filter((r) => this.env.model.getters.isRangeValid(r.xc));
     if (this.props.required && ranges.length === 0) {
       this.state.isMissing = true;
     }
+    for (const range of ranges) {
+      const zone = toUnboundedZone(range.xc);
+      if (zone.bottom !== zone.top && zone.left != zone.right) {
+        if (zone.right) {
+          //test if we don't have something like 1:2
+          this.env.model.dispatch("CHANGE_RANGE", {
+            id: this.id,
+            rangeId: range.id,
+            value: zoneToXc({
+              left: zone.left,
+              right: zone.left,
+              top: zone.top,
+              bottom: zone.bottom,
+            }),
+          });
+          for (var j = zone.left + 1; j <= zone.right; ++j) {
+            this.env.model.dispatch("ADD_RANGE", {
+              id: this.id,
+              value: zoneToXc({
+                left: j,
+                right: j,
+                top: zone.top,
+                bottom: zone.bottom,
+              }),
+            });
+          }
+        } else {
+          this.env.model.dispatch("CHANGE_RANGE", {
+            id: this.id,
+            rangeId: range.id,
+            value: zoneToXc({
+              left: zone.left,
+              right: zone.right,
+              top: zone.top,
+              bottom: zone.top,
+            }),
+          });
+          for (var j = zone.top + 1; j <= zone.bottom!; ++j) {
```

Comment:
please never use `var`
[4 Reasons Why ‘var’ is Considered Obsolete in Modern JavaScript](https://javascript.plainenglish.io/4-reasons-why-var-is-considered-obsolete-in-modern-javascript-a30296b5f08f)